### PR TITLE
Run testsuite on PHP 8

### PR DESCRIPTION
**Pull request by ADmad**

### CODE REVIEW

**Comment by dereuromark** [comment]
Do we want to do that for 4.1 instead? Given the typehint change and alike, fitting 4.1+ here might be easier.

**Comment by ADmad** [comment]
The reason I targeted 4.0 is because of [this](https://travis-ci.com/github/cakephp/cakephp/jobs/341552572#L581) error:

```
10) Cake\Test\TestCase\Database\QueryTest::testSQLFunctions
TypeError: abs(): Argument #1 ($number) must be of type int|float, string given

/home/travis/build/cakephp/cakephp/tests/TestCase/Database/QueryTest.php:3595
/home/travis/build/cakephp/cakephp/tests/TestCase/DatabaseSuite.php:72
```

https://github.com/cakephp/cakephp/blob/a98ed4ef6a55cb524d67da33e668fbde76edb78f/tests/TestCase/Database/QueryTest.php#L3591-L3595

I debugged locally and the value of `$resut[0]['d']` is `3.0`. I am wondering why is it a string when the return type for the expression is specified as `'integer'`?

https://github.com/cakephp/cakephp/blob/a98ed4ef6a55cb524d67da33e668fbde76edb78f/src/Database/FunctionsBuilder.php#L180-L183

Am I miss understanding what the `$return` argument means here:
https://github.com/cakephp/cakephp/blob/a98ed4ef6a55cb524d67da33e668fbde76edb78f/src/Database/FunctionsBuilder.php#L39-L46

**Comment by ADmad** [comment]
The typehint changes would target `4.1` but can fix any other errors like the one above (if it actually is an issue) for `4.0` itself.

File: src/Utility/Hash.php [hunk 1]
```diff
@@ -792,16 +792,18 @@ protected static function _merge(array $stack, array &$return): void
         while (!empty($stack)) {
             foreach ($stack as $curKey => &$curMerge) {
                 foreach ($curMerge[0] as $key => &$val) {
-                    $isArray = is_array($curMerge[1]);
+                    if (!is_array($curMerge[1])) {
```
[the remark this answers is not in the document]
Comment (othercorey) [reply src/Utility/Hash.php:795]: I assume it was hitting the else condition below without this? 
`$curMerge[1][$key] = $val;`

[the remark this answers is not in the document]
Comment (ADmad) [reply src/Utility/Hash.php:795]: Yup

**Comment by ADmad** [comment]
All errors on PHP 8 expect the one mentioned in https://github.com/cakephp/cakephp/pull/14644#issuecomment-636142834 are now addressed.

The other remaining error for `ArrayIterator::ksort()` argument count seems to be a bug in PHP 8.

**Comment by othercorey** [comment]
> The other remaining error for `ArrayIterator::ksort()` argument count seems to be a bug in PHP 8.

Should we open a bug report?

File: src/Utility/Hash.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (othercorey) [reply src/Utility/Hash.php:795]: I wonder if we should fix this in 3.x and add an explicit test for that pattern.

[the remark this answers is not in the document]
Comment (ADmad) [reply src/Utility/Hash.php:795]: We could fix it for 3.x but AFAIK we haven't had any actual errors reported due to this and I don't really see us providing PHP 8 support for 3.x.

**Comment by ADmad** [comment]
> Should we open a bug report?

https://bugs.php.net/bug.php?id=79653

**Comment by ADmad** [comment]
@markstory Need your input regarding https://github.com/cakephp/cakephp/pull/14644#issuecomment-636142834.

**Comment by markstory** [comment]
I agree that dateDiff() should be casting it's results to an integer given that code. I'm not sure why it is failing to do that right now.

**Comment by ADmad** [comment]
So the values doesn't cast to string is because `Database\Query::sql()` never populates it's `Typemap`. Only `ORM\Query::sql()` populates the type map through `_addDefaultSelectTypes()` by iterating the select clause.

This makes me wonder why is `$typeCastEnabled` a property of `Database\Query` instead of `ORM\Query`.

**Comment by ADmad** [comment]
https://bugs.php.net/bug.php?id=79653 is fixed as should be available on travis soon.

So this is good to merge.

**Comment by othercorey** [comment]
> Dealing with this is beyond the scope of this PR so I have just updated the assertion instead to avoid error on PHP 8.

Should we leave it failing so we don't lose this or create a separate tracking issue?

**Comment by ADmad** [comment]
> Should we leave it failing so we don't lose this or create a separate tracking issue?

It should be a separate issue.

**Pull request merged**